stdlib/LinearAlgebra/src/dense.jl

```diff
@@ -634,17 +633,21 @@ function exp!(A::StridedMatrix{T}) where T<:BlasFloat
             C = T[120.,60.,12.,1.]
         end
         A2 = A * A
-        P  = copy(Inn)
-        U  = C[2] * P
-        V  = C[1] * P
-        for k in 1:(div(size(C, 1), 2) - 1)
+        # Compute U and V: Even/odd terms in Padé numerator & denom
+        # Expansion of k=1 in for loop
+        P = A2
+        U = C[2]*I + C[4]*P
+        V = C[1]*I + C[3]*P
+        for k in 2:(div(size(C, 1), 2) - 1)
             k2 = 2 * k
             P *= A2
             U += C[k2 + 2] * P
             V += C[k2 + 1] * P
```

Comment:
Is this recognized as being implementable in-place, or should we enforce it via:
```suggestion
            mul!(U, C[k2 + 2], P, true, true)
            mul!(V, C[k2 + 1], P, true, true)
```
?

Reply:
It does lead to less allocations, thanks. 
```
function f1(A,B,c)
    for j=1:10
        B +=A*c
    end
end
function f2(A,B,c)
    for j=1:10
        mul!(A,c,B,true,true)
    end
end
```
```
julia> A=randn(100,100); B=randn(100,100);
julia> @btime f1($A,$B,0.3)
  120.343 μs (40 allocations: 1.53 MiB)
julia> @btime f2($A,$B,0.3)
  59.444 μs (0 allocations: 0 bytes)
```

I can change that, sure. (Although that actually comes from the old code.) 

Is the five-argument `mul!` always supported for a scalar second argument? I don't see it in the documentation:

![Screenshot_20210504_094802](https://user-images.githubusercontent.com/11163595/116974080-d5299280-acbd-11eb-8f84-1b2ad7228881.png)

Reply:
> Is the five-argument `mul!` always supported for a scalar second argument?

I think so.
```julia
julia> @which mul!(rand(3,3), rand(), rand(3,3), true, true)
mul!(C::AbstractArray, s::Number, X::AbstractArray, alpha::Number, beta::Number) in LinearAlgebra at /Applications/Julia-1.6.app/Contents/Resources/julia/share/julia/stdlib/v1.6/LinearAlgebra/src/generic.jl:132
```
It may fall back to broadcasting anyway, but it will make sure (as good as it can, depending on the matrix types) that there are no intermediate memory allocations.

Reply:
With in-place operations one needs to be sure that, in the course of the process, the eltype cannot be widened. But it doesn't look like that all of sudden, one may need higher accuracy or a complex eltype or something like that, right?

Reply:
By construction we have `eltype(P)=eltype(U)=eltype(V)=eltype(C)=T` so that should be fine.

Reply:
Does this further improve performance?

Reply:
Yes. 
```
julia> A=randn(50,50); A=2.0*A/opnorm(A,1);
julia> @btime exp!(copy(A)); # Original
  477.251 μs (61 allocations: 570.70 KiB)
julia> @btime exp_new!(copy(A)); # Original PR version
  422.929 μs (51 allocations: 472.50 KiB)
julia> @btime exp_new2!(copy(A)); # After @dkarrasch suggestion
  409.678 μs (27 allocations: 236.81 KiB)
```
The improvement is less dramatic for larger matrices, and also for smaller norm. Certainly worthwhile.